src/handle.rs

```diff
@@ -419,3 +437,287 @@ impl HandleHost {
+///
+/// Note that finalization callbacks are tied to the lifetime of a `Weak<T>`,
+/// and will not be called after the `Weak<T>` is dropped.
+///
+/// # `PartialEq` and `Hash`
+///
+/// The implementations of [`PartialEq`] and [`Hash`] might change during the
+/// lifetime of a `Weak<T>` instance: when compared via equals, a non-empty weak
+/// reference will not be equal to a different non-empty weak reference, but it
+/// will once both objects have been GC'd. Likewise, the hash of a weak
+/// reference will change after GC. That makes `Weak<T>` not suitable for use as
+/// the key for a [`HashMap`] or [`HashSet`].
+///
+/// # `Clone`
+///
+/// Since finalization callbacks are specific to a `Weak<T>` instance, cloning
+/// will create a new object reference without a finalizer, as if created by
+/// [`Self::new`]. You can use [`Self::clone_with_finalizer`] to attach a
+/// finalization callback to the clone.
+pub struct Weak<T> {
+  data: Option<Pin<Box<WeakData<T>>>>,
+  isolate_handle: IsolateHandle,
+}
+
+impl<T> Weak<T> {
+  pub fn new(isolate: &mut Isolate, handle: impl Handle<Data = T>) -> Self {
+    let HandleInfo { data, host } = handle.get_handle_info();
+    host.assert_match_isolate(isolate);
+    Self::new_raw(isolate, data, None)
+  }
+
+  /// Create a Weak handle with a finalization callback installed.
+  ///
+  /// NOTE: There is no guarantee as to *when* or even *if* the callback is
+  /// invoked. The invocation is performed solely on a best effort basis. As
+  /// always, GC-based finalization should *not* be relied upon for any critical
+  /// form of resource management!
+  ///
+  /// The finalizer does not have access to the inner value, because it has
+  /// already been GC'd by the time it runs.
+  pub fn with_finalizer(
+    isolate: &mut Isolate,
+    handle: impl Handle<Data = T>,
+    finalizer: Box<dyn FnOnce()>,
+  ) -> Self {
+    let HandleInfo { data, host } = handle.get_handle_info();
+    host.assert_match_isolate(isolate);
+    Self::new_raw(isolate, data, Some(finalizer))
+  }
+
+  fn new_raw(
+    isolate: *mut Isolate,
+    data: NonNull<T>,
+    finalizer: Option<Box<dyn FnOnce()>>,
+  ) -> Self {
+    let weak_data = Box::pin(WeakData {
+      pointer: Default::default(),
+      finalizer: Cell::new(finalizer),
+    });
+    let data = data.cast().as_ptr();
+    let data = unsafe {
+      v8__Global__NewWeak(
+        isolate,
+        data,
+        &weak_data as *const _ as *const c_void,
+        Self::first_pass_callback,
+      )
+    };
+    weak_data
+      .pointer
+      .set(Some(unsafe { NonNull::new_unchecked(data as *mut _) }));
+    Self {
+      data: Some(weak_data),
+      isolate_handle: unsafe { (*isolate).thread_safe_handle() },
+    }
+  }
+
+  pub fn empty(isolate: &mut Isolate) -> Self {
+    Weak {
+      data: None,
+      isolate_handle: isolate.thread_safe_handle(),
+    }
+  }
+
+  pub fn clone_with_finalizer(&self, finalizer: Box<dyn FnOnce()>) -> Self {
```

Comment:
`Weak<T>` doesn't implement `Handle`, and changing `Handle` to support weak references is probably more trouble than it's worth.

Reply:
> Weak<T> doesn't implement Handle

Oh, right.

> because changing Handle to support weak references is probably more trouble than it's worth.

I agree with that!